### PR TITLE
Fix persistent worker exits before pin_memory thread

### LINKED ISSUE #1551: Multithreaded DataLoader sometimes hits "Connection reset by peer"
I am using `torch.utils.data.DataLoader` with `num_workers = 4` and sometimes getting this exception (in a single-threaded mode it works fine). I will try to get minimal repro.

I suppose it may be related to https://discuss.pytorch.org/t/using-torch-tensor-over-multiprocessing-queue-process-fails/2847/3 but in this case it's not custom Queue handling.

```
Traceback (most recent call last):
........
  File ".../torch/utils/data/dataloader.py", line 206, in __next__
    idx, batch = self.data_queue.get()
  File "/usr/lib/python2.7/multiprocessing/queues.py", line 378, in get
    return recv()
  File ".../torch/multiprocessing/queue.py", line 22, in recv
    return pickle.loads(buf)
  File "/usr/lib/python2.7/pickle.py", line 1388, in loads
    return Unpickler(file).load()
  File "/usr/lib/python2.7/pickle.py", line 864, in load
    dispatch[key](self)
  File "/usr/lib/python2.7/pickle.py", line 1139, in load_reduce
    value = func(*args)
  File ".../torch/multiprocessing/reductions.py", line 68, in rebuild_storage_fd
    fd = multiprocessing.reduction.rebuild_handle(df)
  File "/usr/lib/python2.7/multiprocessing/reduction.py", line 155, in rebuild_handle
    conn = Client(address, authkey=current_process().authkey)
  File "/usr/lib/python2.7/multiprocessing/connection.py", line 175, in Client
    answer_challenge(c, authkey)
  File "/usr/lib/python2.7/multiprocessing/connection.py", line 432, in answer_challenge
    message = connection.recv_bytes(256)         # reject large message
IOError: [Errno 104] Connection reset by peer
```

cc @SsnL @VitalyFedyunin @ejguan

### PR DESCRIPTION
**Pull request by ejguan**
Fixes #1551

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#71579 Fix persistent worker exits before pin_memory thread**

As the comment in the code, register a function to terminate persistent workers. 
By adding a reference of these workers in `atexit`, it would prevent Python interpreter kills these persistent worker processes before `pin_memorh_thread` exits.
And, if users explicitly kills DataLoader iterator, such function in `atexit` would be a no-op.

Differential Revision: [D33896537](https://our.internmc.facebook.com/intern/diff/D33896537)

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -946,6 +946,18 @@ def __init__(self, loader):
         else:
             self._data_queue = self._worker_result_queue
 
+        # In some rare cases, persistent workers (daemonic processes)
+        # would be terminated before `__del__` of iterator is invoked
+        # when main process exits
+        # It would cause failure when pin_memory_thread tries to read
+        # corrupted data from worker_result_queue
+        # atexit is used to prevent persistent workers exiting before
+        # pin_memory_thread
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:955]: Here is the comment.

File: test/test_dataloader.py [hunk 2]
```diff
@@ -2302,6 +2302,41 @@ def test_dataset_not_reset(self):
...
+        batch_size=16,
+        num_workers=2,
+        pin_memory=True,
+        persistent_workers=True,
+    )
+
+    for _ in dl:
+        break
+"""], stderr=subprocess.PIPE)
+        self.assertTrue(proc.stderr == b'')
```
Comment (ejguan) [inline test/test_dataloader.py:2339]: This test won't pass before this PR.

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694867).

File: torch/utils/data/dataloader.py [hunk 3]
```diff
@@ -946,6 +946,18 @@ def __init__(self, loader):
...
+        if self._persistent_workers and self._pin_memory:
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:956]: In order to potentially fix part of #71187 for the case of persistent workers, we may want to remove this condition on `self._pin_memory`

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by malfet** [comment]
Reverting as it broke Windows test sanity(from https://github.com/pytorch/pytorch/runs/4903195822?check_suite_focus=true):
```
Traceback (most recent call last):
  File "test_dataloader.py", line 2339, in test_early_exit
    self.assertTrue(proc.stderr == b'')
AssertionError: False is not true```

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33896537).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33896537).

**Comment by anjali411** [comment]
@ejguan can you also put a topic label to it? 😊

**Pull request closed without merging**